proxy/src/main/java/com/velocitypowered/proxy/connection/client/ConnectedPlayer.java

```diff
@@ -553,7 +555,15 @@ void teardown() {
       connectedServer.disconnect();
     }
     server.unregisterConnection(this);
-    server.getEventManager().fireAndForget(new DisconnectEvent(this));
+    server.getEventManager().fire(new DisconnectEvent(this))
+            .thenRun(() -> this.teardownFuture.complete(null));
+  }
+
+  public void awaitTeardown() {
+    try {
+      teardownFuture.get();
```

Comment:
Maybe a timeout? In case a plugin takes too much time to handle the event.

Reply:
There should be a timeout. Further, it needs to be over all player disconnects, not just individual player disconnects.